_posts/2020-02-07-metal3-dev-env-install-deep-dive.md

```diff
@@ -0,0 +1,921 @@
+deployment.apps/metal3-baremetal-operator created
+```
+
+Once the `BMO` objects are applied, it's time to transform the virtual bare metal hosts information into a yaml file of kind `BareMetalHost` Custom Resource (CR). This is done by a golang script passing them the IPMI address, BMC username and password, which are stored as a Kubernetes secret, MAC address and name:
+
+```sh
++ go run /home/alosadag/go/src/github.com/metal3-io/baremetal-operator/cmd/make-bm-worker/main.go -address ipmi://192.168.111.1:6230 -password password -user admin -boot-mac 00:be:bc:fd:17:f3 node-0
++ read -r name address user password mac
++ go run /home/alosadag/go/src/github.com/metal3-io/baremetal-operator/cmd/make-bm-worker/main.go -address ipmi://192.168.111.1:6231 -password password -user admin -boot-mac 00:be:bc:fd:17:f7 node-1
++ read -r name address user password mac
++ go run /home/alosadag/go/src/github.com/metal3-io/baremetal-operator/cmd/make-bm-worker/main.go -address ipmi://192.168.111.1:6232 -password password -user admin -boot-mac 00:be:bc:fd:17:fb node-2
++ read -r name address user password mac
+```
+<br>
+Below is shown the bare metal host definition of node-1. Note that the IPMI address is the IP of the host's provisioning interface. Behind the scenes, IPMI is handled by the `vbmc` container running in the host.
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: node-1-bmc-secret
+type: Opaque
+data:
+  username: YWRtaW4=
+  password: cGFzc3dvcmQ=
+---
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: node-1
+spec:
+  online: true
+  bootMACAddress: 00:00:e0:4b:24:8f
+  bmc:
+    address: ipmi://192.168.111.1:6231
+    credentialsName: node-1-bmc-secret
+```
+
+See that the MAC address configured in the `BareMetalHost` spec definition matches *node-1* provisioning interface:
+
+```sh
+[root@eko1 metal3-dev-env]# virsh domiflist node_1
+Interface  Type       Source     Model       MAC
+-------------------------------------------------------
+vnet4      bridge     provisioning virtio      00:00:e0:4b:24:8f
+vnet5      bridge     baremetal  virtio      00:00:e0:4b:24:91
+```
+
+Finally, the script apply in namespace metal3 each of the `BareMetalHost` yaml files that match each virtual bare metal host:
+
+```sh
++ kubectl apply -f bmhosts_crs.yaml -n metal3
+secret/node-0-bmc-secret created
+baremetalhost.metal3.io/node-0 created
+secret/node-1-bmc-secret created
+baremetalhost.metal3.io/node-1 created
+secret/node-2-bmc-secret created
+baremetalhost.metal3.io/node-2 created
+```
+
+Lastly, it is the turn of the `CAPBM`. Similar to `BMO`, `kustomize` is used to create the different Kubernetes components and `kubectl` applied the files into the management cluster.
+
+<br>
+> **Warning:** Note that installing `CAPBM` includes installing the components of the [Cluster API](https://github.com/kubernetes-sigs/cluster-api) and the components of the `Cluster API bootstrap provider kubeadm` (CABPK)
```

Comment:
Probalby also link the [CABPK](https://github.com/kubernetes-sigs/cluster-api/tree/master/bootstrap/kubeadm)